src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -202,22 +202,35 @@ private function registerBusToCollector(ContainerBuilder $container, string $bus
 
     private function registerBusMiddleware(ContainerBuilder $container, string $busId, array $middleware)
     {
-        $container->getDefinition($busId)->replaceArgument(0, array_map(function (string $name) use ($container, $busId) {
-            if (!$container->has($messengerMiddlewareId = 'messenger.middleware.'.$name)) {
-                $messengerMiddlewareId = $name;
+        $middlewareReferences = array();
+        foreach ($middleware as $middlewareItem) {
+            $id = $middlewareItem['id'];
+            $arguments = $middlewareItem['arguments'] ?? array();
+            if (!$container->has($messengerMiddlewareId = 'messenger.middleware.'.$id)) {
+                $messengerMiddlewareId = $id;
             }
 
             if (!$container->has($messengerMiddlewareId)) {
-                throw new RuntimeException(sprintf('Invalid middleware "%s": define such service to be able to use it.', $name));
+                throw new RuntimeException(sprintf('Invalid middleware "%s": define such service to be able to use it.', $id));
             }
 
-            if ($container->getDefinition($messengerMiddlewareId)->isAbstract()) {
+            if (($definition = $container->findDefinition($messengerMiddlewareId))->isAbstract()) {
                 $childDefinition = new ChildDefinition($messengerMiddlewareId);
+                $count = \count($definition->getArguments());
+                foreach (array_values($arguments ?? array()) as $key => $argument) {
+                    // Parent definition can provide default arguments.
+                    // Replace each explicitly or add if not set:
+                    $key < $count ? $childDefinition->replaceArgument($key, $argument) : $childDefinition->addArgument($argument);
```

Comment:
Why not just using `setArgument` here?

Reply:
Due to the way child definitions arguments work, using `replaceArgument` saves an extra hard coded `index_` prefix for the key here.

https://github.com/symfony/symfony/blob/d264fcea24cc279054c7fe480e3dafb67f694668/src/Symfony/Component/DependencyInjection/ChildDefinition.php#L98-L99